src/transformStateWithClones.js

```diff
@@ -7,7 +7,31 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const states = [];
+  let currentState = { ...state };
+
+  actions.forEach((action) => {
+    switch (action.type) {
+      case 'clear':
+        currentState = {};
+        break;
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+      case 'removeProperties':
+        if (Array.isArray(action.keysToRemove)) {
+          action.keysToRemove.forEach((key) => {
+            delete currentState[key];
+          });
+        }
+        break;
+      default:
+        break;
+    }
+    states.push({ ...currentState });
```

Comment:
You're correctly avoiding direct mutations to the `currentState` object within the switch cases, but the action of copying the `currentState` to the `states` array is repeated in every case. To adhere to the DRY principle, consider moving this action outside of the switch statement, so it's done once after the switch block, for each iteration of the `forEach` loop.

Reply:
Still not fixed.